File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/agent/GoalExecutor.kt
package org.wyrdsekai.app.engine.agent
import org.wyrdsekai.app.platform.formatFixed

/**
 * Decision engine for goal-based task execution (KMP port).
 * Pure functions — no state. Mirrors Java GoalExecutor.
 */
object GoalExecutor {

    private const val MAX_SAME_ACTION_REPEATS = 3
    private const val MAX_TOTAL_PLAN_ATTEMPTS = 15
    private const val ENERGY_FLOOR = 0.15

    sealed interface Decision {
        data class Advance(val outcome: String) : Decision
        data class Retry(val guidance: String) : Decision
        data class Delegate(val targetAgent: String, val task: String) : Decision
        data class Replan(val reason: String) : Decision
        data class Skip(val reason: String) : Decision
        data class Escalate(val message: String) : Decision
        data class Suspend(val reason: String) : Decision
    }

    fun evaluate(
        plan: TaskPlan,
        actionSuccess: Boolean,
        actionResult: String?,
        actionType: String,
        currentEnergy: Double
    ): Decision {
        val goal = plan.currentGoal() ?: return Decision.Advance("No active goal")

        if (actionSuccess) {
            return Decision.Advance(actionResult ?: "completed")
        }

        // Energy floor
        if (currentEnergy < ENERGY_FLOOR) {
            return Decision.Suspend("Energy too low (${formatFixed(currentEnergy, 2)}) — will resume after rest")
        }

        // Total attempts exceeded
        if (plan.totalAttempts >= MAX_TOTAL_PLAN_ATTEMPTS) {
            return Decision.Escalate("Exhausted $MAX_TOTAL_PLAN_ATTEMPTS total attempts. Last: $actionResult")
        }

        // Can retry?
        if (goal.hasRetriesLeft()) {
            if (detectRepeatLoop(goal, actionType)) {
                return Decision.Retry("STOP repeating '$actionType'. Try a completely different approach.")
            }
            val remaining = goal.attemptsRemaining()
            val guidance = if (goal.attempts.size == 1)
                "Try a different approach ($remaining retries remaining)."
            else
                "Previous approaches haven't worked. Try completely different strategy ($remaining retries remaining)."
            return Decision.Retry("$actionType failed: $actionResult. $guidance")
        }

        // Can skip?
        if (plan.currentGoalIndex < plan.goals.size - 1) {
            return Decision.Skip("Failed after ${goal.attempts.size} attempts: $actionResult")
        }

        // Escalate
        val tried = goal.attempts.map { "${it.actionType}: ${it.result ?: "failed"}" }
        return Decision.Escalate("Stuck on '${goal.description}'. Tried: $tried")
    }

    private fun detectRepeatLoop(goal: Goal, actionType: String): Boolean {
        val attempts = goal.attempts
        if (attempts.size < MAX_SAME_ACTION_REPEATS) return false
        var count = 0
        for (i in attempts.indices.reversed()) {
            if (attempts[i].actionType == actionType) count++ else break
            if (count >= MAX_SAME_ACTION_REPEATS) return true
        }
        return false
    }
}

File: clients/rn/src/engine/agent/GoalExecutor.ts
/**
 * Decision engine for goal-based task execution (RN port).
 * Pure functions. Mirrors Java GoalExecutor.
 */

import {TaskPlan, Goal, currentGoal} from './TaskPlan';

const MAX_SAME_ACTION_REPEATS = 3;
const MAX_TOTAL_PLAN_ATTEMPTS = 15;
const ENERGY_FLOOR = 0.15;

export type Decision =
  | {type: 'advance'; outcome: string}
  | {type: 'retry'; guidance: string}
  | {type: 'delegate'; targetAgent: string; task: string}
  | {type: 'replan'; reason: string}
  | {type: 'skip'; reason: string}
  | {type: 'escalate'; message: string}
  | {type: 'suspend'; reason: string};

export function evaluate(
  plan: TaskPlan,
  actionSuccess: boolean,
  actionResult: string | undefined,
  actionType: string,
  currentEnergy: number,
): Decision {
  const goal = currentGoal(plan);
  if (!goal) return {type: 'advance', outcome: 'No active goal'};

  if (actionSuccess) {
    return {type: 'advance', outcome: actionResult ?? 'completed'};
  }

  if (currentEnergy < ENERGY_FLOOR) {
    return {type: 'suspend', reason: `Energy too low (${currentEnergy.toFixed(2)})`};
  }

  if (plan.totalAttempts >= MAX_TOTAL_PLAN_ATTEMPTS) {
    return {type: 'escalate', message: `Exhausted ${MAX_TOTAL_PLAN_ATTEMPTS} attempts. Last: ${actionResult}`};
  }

  const remaining = Math.max(0, goal.maxAttempts - goal.attempts.length);
  if (remaining > 0) {
    if (detectRepeatLoop(goal, actionType)) {
      return {type: 'retry', guidance: `STOP repeating '${actionType}'. Try completely different approach.`};
    }
    const hint = goal.attempts.length === 1
      ? 'Try a different approach'
      : 'Previous approaches failed. Try completely different strategy';
    return {type: 'retry', guidance: `${actionType} failed: ${actionResult}. ${hint} (${remaining} retries).`};
  }

  if (plan.currentGoalIndex < plan.goals.length - 1) {
    return {type: 'skip', reason: `Failed after ${goal.attempts.length} attempts: ${actionResult}`};
  }

  const tried = goal.attempts.map(a => `${a.actionType}: ${a.result ?? 'failed'}`);
  return {type: 'escalate', message: `Stuck on '${goal.description}'. Tried: ${tried.join(', ')}`};
}

function detectRepeatLoop(goal: Goal, actionType: string): boolean {
  const attempts = goal.attempts;
  if (attempts.length < MAX_SAME_ACTION_REPEATS) return false;
  let count = 0;
  for (let i = attempts.length - 1; i >= 0; i--) {
    if (attempts[i].actionType === actionType) count++;
    else break;
    if (count >= MAX_SAME_ACTION_REPEATS) return true;
  }
  return false;
}
